File: solutions/fortran/allergies/1/allergies.f90

module allergies
  implicit none
character(len=12), dimension(8), parameter :: allergens = & 
[character(len=12) :: 'eggs', 'peanuts', 'shellfish', &
'strawberries', 'tomatoes', 'chocolate', 'pollen', 'cats']

contains

  logical function allergicTo(allergy_str, allergy_key) result(res)
    character(len=*), intent(in) :: allergy_str
    integer, intent(in) :: allergy_key
    res = (index(allergicList(allergy_key), allergy_str) /= 0) 
  end function



  function allergicList(allergy_key)
    integer, intent(in) :: allergy_key
    character(len=100) :: allergicList
    integer :: i
    allergicList = ' '

    do i = 1,8
      if (iand(ishft(allergy_key, -i+1),1).eq.1) then
        if (len(trim(allergicList)).eq.0) then
          allergicList =  trim(allergens(i))
        else
          allergicList =  trim(allergicList) //" "// trim(allergens(i))
        end if
      end if
    end do
  end function



end module

File: solutions/dart/allergies/1/lib/allergies.dart
import 'package:collection/collection.dart';

class Allergies {   

  static final Map<String,int> allergens = 
 ['eggs', 'peanuts', 'shellfish', 'strawberries', 'tomatoes', 'chocolate', 'pollen', 'cats'].asMap().map((i,e)=> MapEntry(e, 1<<i));
    
  bool allergicTo(String name, int score){
    return ((allergens[name] ?? 0) & score) != 0;    
  }

  List<String> list(int score){
    return allergens.keys.where((k)=> allergicTo(k,score)).toList();
  }
}
